engine/entity.py

```diff
@@ -1,81 +1,131 @@
-from copy import deepcopy
-
-class State:
-  def __init__(self, entity_id, raw_state={}):
-    self.entity_id = entity_id
-    self.raw_state = raw_state
-  
-  def __contains__(self, key):
-    return key in self.raw_state
-
-  def get(self, key):
-    return deepcopy(self.raw_state[key]) if key in self.raw_state else None
-  
-  def set(self, key, value, diffs={}):
-    new_diffs = deepcopy(diffs)
- 
-    if issubclass(value.__class__, Entity):
-      print('Warning: You attempted to set state: {key} {value} Do not set entities to state. Defaulting to id.')
-      value = value.id
-    
-    original_value = self.get(key)
-    self.raw_state[key] = value
-
-    if self.entity_id not in new_diffs:
-      new_diffs[self.entity_id] = {}
-    
-    new_diffs[self.entity_id][key] = (original_value, value)
-
-    return new_diffs
+from engine.diff import Diff
+from engine.state import State
 
 class Entity:
   next_entity_id = 0
 
-  def __init__(self, children=None, parent=None, reactions=None, state=None):
-    self.children = {}
-    self.does_exist = True
-    self.reactions = reactions if reactions is not None else self.get_default_reactions()
-    
+  def __init__(self, children=[], parent=None, state={}):
+    self.descendants = {}
+    self.diffs = []
+    self.listeners = {}
+    self.parent = None
+    self.root = self
+
     self.id = Entity.next_entity_id
     Entity.next_entity_id += 1
-
-    raw_state = state if state is not None else self.get_default_state()
-    self.state = State(self.id, raw_state)
-
-    children = children if children is not None else self.get_default_children()
-    
-    for child in children:
-      self.add_child(child)
     
     if parent is not None:
       parent.add_child(self)
+    
+    self.children = {}
+    children_list = self.get_default_children() + children
 
-  def add_child(self, child):
-    self.children[child.id] = child
-    child.parent = self
+    for child in children_list:
+      self.add_child(child)
 
-  def get_default_children(self):
-    return []
+    raw_state = self.get_default_state()
+
+    for key,value in state.items():
+      raw_state[key] = value
+
+    self.state = State(raw_state)
+
+  def add_child(self, child, diff=Diff()):
+    if child.id not in self.children:
+      if child.parent is not None:
+        child.parent.remove_child(child)
 
-  def get_default_reactions(self):
+      self.children[child.id] = child
+      child.parent = self
+      child.root = self.root
+      self.root.descendants[child.id] = child
+
+      if hasattr(child, 'get_should_react'):
+        self.root.listeners[child.id] = child
+
+      for descendant_id in child.descendants:
+        descendant = child.descendants[descendant_id]
+        self.root.descendants[descendant_id] = descendant
+        descendant.root = self.root
+
+      for listener_id in child.listeners:
+        self.root.listeners[listener_id] = child.listeners[listener_id]
+
+      child.descendants = {}
```

Comment:
@nigelrobbins3 in order to minimize unnecessary storage. The current intent is that you only need the root entity to keep the descendant registry. I added a hydrate method that you should be able to use instead of having to access the descendants registry directly, and it uses root.

Reply:
:+1: